### ISSUE TITLE
Improve Jest availability detection by project

### ISSUE THREAD

**Issue by rfgamaral** [jest-community/vscode-jest#479]
Right now this extension checks for a `jest.config.js` or `jest.json` and/or if `node_modules/.bin/jest` is installed. I propose a few improvements to this because the actual behavior is that this extension still attempts to run the Jest process when it can't.

The current behavior, if I'm not mistaken, is:

1. Do we have a `jest.config.js` or `jest.json` and `node_modules/.bin/jest` exists? Run `node_modules/.bin/jest`.
2. Do we have a `jest.config.js` or `jest.json` but `node_modules/.bin/jest` doesn't exist? Run a supposedly installed global `jest`.
3. There's no `jest.config.js` or `jest.json` and `node_modules/.bin/jest` exists? Run `node_modules/.bin/jest`.
4. There's no `jest.config.js` or `jest.json` but `node_modules/.bin/jest` doesn't exist? Run a supposedly installed global `jest`.
5. There's no `jest.config.js` or `jest.json` and `node_modules/.bin/jest` doesn't exist? Don't do anything.

**About 1 and 3)**

I know Microsoft is pushing for the Remote WSL extension and deprecating support for everything without that extension. That's fine. But sometimes I still open a project, with all it's dependencies installed in WSL, as a Windows folder for some quick edits. In this situation, although `node_modules/.bin/jest` will be found, you will be trying to run a Linux command inside Windows, it will fail:

```
'c:\Users\Ricardo\Workspace\Project\node_modules\.bin\jest' is not recognized as an internal or external command, operable program or batch file.

Starting Jest in Watch mode failed too many times and has been stopped.
 see troubleshooting: https://github.com/jest-community/vscode-jest/blob/master/README.md#troubleshooting
```

My first suggestion is, check if `jest` is a binary you can run, if not, don't even attempt to call it. This avoids these output messages and this error notification:

![image](https://user-images.githubusercontent.com/96476/60759548-ccfb6000-a01e-11e9-8dcc-8d73b8282f3b.png)

**About 2 and 4)**

Assuming it's intended behavior to run a global `jest` if one is not found in `node_modules`, I think the same suggestion above applies here because I get similar errors:

![image](https://user-images.githubusercontent.com/96476/60759568-206dae00-a01f-11e9-84b6-bceabe8b4c02.png)

In other words, this specific project doesn't have `node_modules/.bin/jest`, nor `jest` is available globally in my system, so why is this extension still trying to run it? It shouldn't, right?

<hr>

I believe all scenarios could be easily fixed, even for projects inside WSL opened without the Remote WSL extension, the solution should just work for every scenario. I mean, a simple "can I run this command (global or node_modules, doesn't matter) or will I get an error"? If I can, run it, otherwise, don't (just like point 5, which works perfectly).

Is this something that could be considered?

**Comment by connectdotz** [comment]
the current behavior is actually quite simple, there are mainly 2 parts:

first, the extension is bootstrapped with vscode's [activation events](https://code.visualstudio.com/api/references/activation-events), which only supports static configuration today. You can see vscode-jest's activation events [here](https://github.com/jest-community/vscode-jest/blob/07bc3e0947dc498bd5f5ddffed9882bf7fb2e6c8/package.json#L24).

second, once the extension is activated, it looks for custom `jest.pathToJest` to start jest, if not found, it will then fall back to this [default pattern](https://github.com/jest-community/vscode-jest/blob/6d0ea5d19c84dd9d5fa4695ca65d9766dd4c6a9e/src/helpers.ts#L64) to start jest. 

(note: while the default pattern would work for most simple/standard projects, it is most likely not adequate for more sophisticated projects, which is why there are settings like `jest.pathToJest`, `jest.pathToConfig` to help users customize.)

A jest process can fail for many reasons, such as missing (being hoisted, haven't installed), bad symlink, incompatible binary format, incorrect jest config, or even invalid nodejs, just to name a few. Fortunately, there is no need to test for each possible condition because we can simply run it to find out. 

The fact you saw error notification is precisely due to it detected the jest process failure. This is not uncommon and as mentioned above, users often either fix the error or customize the extension. Until the tests can be run, all tests will be marked "unknown" (empty circle).

I am not quite sure what you mean by _"If I can, run it, otherwise, don't"_... you mean not to activate the extension? fail silently? or something else?

**Comment by rfgamaral** [comment]
In a nutshell, I don't want to be annoyed (with error notifications, `output` pane popping out, codelens or visual circles) if Jest can't be executed successfully, assuming default options.

After looking at the [default pattern](https://github.com/jest-community/vscode-jest/blob/6d0ea5d19c84dd9d5fa4695ca65d9766dd4c6a9e/src/helpers.ts#L64) to start jest I see room for improvement. I'll open a PR with my suggestions and we go from there, it will be easier for both of us.

Just give me some time :)

**Comment by rfgamaral** [comment]
@connectdotz I've opened a draft PR so you can get that gist of what I'm looking for. Let me know what you think.
